README.md
# MQA URL Checker

## Setup

1. Install Java 11

2. Clone the directory and enter it

        git clone git@gitlab.fokus.fraunhofer.de:viaduct/metrics/metrics-url-checker.git

3. Edit the environment variables in the `Dockerfile` to your liking. Variables and their purpose are listed below:

    |Key|Description|Default|
    |:--- |:---|:---|
    |PORT| The port this service will run on | 8085 |
    |WORKER_COUNT| The number of UrlCheckerVerticles to spawn | 3 |
    |HTTP_USER_AGENT| The user agent to send with the check requests | Mozilla/5.0 (European Data Portal) Gecko/20100101 Firefox/40.1 |
    |TIMEOUT_IN_MILLIS| HTTP request timeout in milliseconds | 10000 |
    |CONNECTION_EXCEPTION_STATUS| The status code to set in case of exceptions | 1100 |
    |INVALID_URL_STATUS| The status code to set if invalid URLs are provided | 1300 |
    |INVALID_URL_MESSAGE| The status message to set if invalid URLs are provided | Invalid URL |

## Run

### Production

Build the project by using the provided Maven wrapper. This ensures everyone this software is provided to can use the exact same version of the maven build tool.
The generated _fat-jar_ can then be found in the `target` directory.

* CLI

        mvn clean package
        java -jar target/application.jar

* Docker

        1. Start your docker daemon
        2. Build the application as described in Windows or Linux
        3. Adjust the port number (`EXPOSE` in the `Dockerfile`)
        4. Build the image: `docker build -t metrics/url-checker .`
        5. Run the image, adjusting the port number as set in step 3: `docker run -i -p 8080:8080 metrics/url-checker`

### Development

## CI

The repository uses the gitlab builtin CI Framework.
The `gitlab-ci.yaml` file contains the appropriate config. With each push a new docker image is built using the `Dockerfile`.
The image is then stored in the gitlab registry, from which it can then be pulled using the following command:

    docker pull registry.gitlab.com/european-data-portal/mqa-url-checker

## API

A formal OpenAPI 3 specification can be found in the `src/main/resources/webroot/openapi.yaml` file.
A visually more appealing version is available at `{url}:{port}` once the application has been started.CHANGELOG.md


CHANGELOG.md
# ChangeLog

## [1.0.0](https://gitlab.fokus.fraunhofer.de/viaduct/metrics/metrics-url-checker/-/tags/1.0.0) (2019-12-05)

**Changed:**
* Bumped Vert.X version to 4.0.0-milestone3
* Changed required Java version to 11
* URLs are checked with a HTTP HEAD instead of GET request

**Removed:**
* Maven Wrapper
* Redeploy scripts

**Fixed:**
* All errors related to the Vert.X version bump


src/test/java/io/piveau/metrics/url_checker/IntegrationTest.java
package io.piveau.metrics.url_checker;

import io.vertx.core.DeploymentOptions;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpServer;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.WebClientOptions;
import io.vertx.ext.web.client.predicate.ResponsePredicate;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.extension.ExtendWith;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.junit.jupiter.api.Assertions.*;

@ExtendWith(VertxExtension.class)
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class IntegrationTest {

    private HttpServer callbackServer;
    private WebClient webClient;

    private static final int TEST_SERVER_PORT = 8081;
    private static final int CALLBACK_PORT = 8082;

    @BeforeAll
    void setup(Vertx vertx, VertxTestContext testContext) {
        WebClientOptions clientOptions = new WebClientOptions()
            .setDefaultHost("localhost")
            .setDefaultPort(ApplicationConfig.DEFAULT_APPLICATION_PORT);
        webClient = WebClient.create(vertx, clientOptions);

        // spawn server offering mock endpoint for testing
        Router router = Router.router(vertx);
        router.head("/valid_url").handler(validRoute ->
            validRoute.response().setStatusCode(200).putHeader("Content-Type", "application/json").end());

        JsonObject testConfig = new JsonObject()
            .put(ApplicationConfig.ENV_APPLICATION_PORT, ApplicationConfig.DEFAULT_APPLICATION_PORT);

        vertx.createHttpServer()
            .requestHandler(router)
            .listen(TEST_SERVER_PORT, testContext.succeeding(mockServer ->
                vertx.deployVerticle(MainVerticle.class.getName(), new DeploymentOptions().setConfig(testConfig), testContext.completing())));
    }

    @AfterEach
    void tearDown(VertxTestContext testContext) {
        callbackServer.close(testContext.completing());
    }

    @Test
    void testReachableUrl(Vertx vertx, VertxTestContext testContext) {
        String testUrl = "http://localhost:" + TEST_SERVER_PORT + "/valid_url";

        JsonObject checkRequest = new JsonObject()
            .put("callback", "http://localhost:" + CALLBACK_PORT + "/valid_url_callback")
            .put("urls", new JsonArray().add(testUrl));

        JsonArray expectedResult = new JsonArray().add(
            new JsonObject()
                .put("url", testUrl)
                .put("statusCode", 200)
                .put("mimeType", "application/json")
                .put("message", "OK"));

        Router router = Router.router(vertx);
        router.post("/valid_url_callback").handler(callback -> {
            testResponse(testContext, callback, expectedResult);
            callback.response().setStatusCode(200).putHeader("Content-Type", "application/json").end();
        });

        // create custom callback server
        callbackServer = vertx.createHttpServer()
            .requestHandler(router)
            .listen(CALLBACK_PORT, testContext.succeeding(callbackServer ->
                webClient.post("/check")
                    .expect(ResponsePredicate.SC_ACCEPTED)
                    .sendJson(checkRequest, testContext.succeeding())));
    }

    @Test
    void testUnreachableUrl(Vertx vertx, VertxTestContext testContext) {
        // this endpoint doesn't exist
        String testUrl = "http://localhost:" + TEST_SERVER_PORT + "/invalid_url";

        JsonObject checkRequest = new JsonObject()
            .put("callback", "http://localhost:" + CALLBACK_PORT + "/invalid_url_callback")
            .put("urls", new JsonArray().add(testUrl));

        JsonArray expectedResult = new JsonArray().add(
            new JsonObject()
                .put("url", testUrl)
                .put("statusCode", 404)
                .putNull("mimeType")
                .put("message", "Not Found"));

        Router router = Router.router(vertx);
        router.post("/invalid_url_callback").handler(callback -> {
            testResponse(testContext, callback, expectedResult);
            callback.response().setStatusCode(200).putHeader("Content-Type", "application/json").end();
        });

        // create custom callback server
        callbackServer = vertx.createHttpServer()
            .requestHandler(router)
            .listen(CALLBACK_PORT, testContext.succeeding(callbackServer ->
                webClient.post("/check")
                    .expect(ResponsePredicate.SC_ACCEPTED)
                    .sendJson(checkRequest, testContext.succeeding())));
    }

    private void testResponse(VertxTestContext testContext, RoutingContext context, JsonArray expectedResult) {
        testContext.verify(() ->
            context.request().bodyHandler(body -> {
                assertNotNull(body);

                try {
                    assertEquals(expectedResult, body.toJsonArray());
                    testContext.completeNow();
                } catch (Exception e) {
                    fail(e);
                }
            }));
    }
}


src/main/java/io/piveau/metrics/url_checker/MainVerticle.java
package io.piveau.metrics.url_checker;

import io.vertx.config.ConfigRetriever;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.DeploymentOptions;
import io.vertx.core.Future;
import io.vertx.core.Promise;
import io.vertx.core.http.HttpServer;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.core.json.DecodeException;
import io.vertx.core.json.Json;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.api.contract.openapi3.OpenAPI3RouterFactory;
import io.vertx.ext.web.handler.StaticHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


public class MainVerticle extends AbstractVerticle {

    private static final Logger log = LoggerFactory.getLogger(MainVerticle.class);

    private JsonObject config;

    @Override
    public void start(Promise<Void> startPromise) {
        log.info("Launching URL-Checker...");

        // startup is only successful if no step failed
        Future<Void> steps = loadConfig()
                .compose(handler -> bootstrapVerticles())
                .compose(handler -> startServer());

        steps.setHandler(handler -> {
            if (handler.succeeded()) {
                log.info("URL-Checker successfully launched");
                startPromise.complete();
            } else {
                log.error("Failed to launch Transformer: " + handler.cause());
                startPromise.fail(handler.cause());
                vertx.close();
            }
        });
    }

    private Future<Void> loadConfig() {
        return Future.future(loadConfig -> {
            ConfigRetriever.create(vertx).getConfig(handler -> {
                if (handler.succeeded()) {
                    config = handler.result();
                    log.debug(config.encodePrettily());
                    loadConfig.complete();
                } else {
                    loadConfig.fail(handler.cause());
                }
            });
        });
    }

    private Future<Void> bootstrapVerticles() {
        DeploymentOptions options = new DeploymentOptions()
                .setInstances(config.getInteger(ApplicationConfig.ENV_WORKER_COUNT, ApplicationConfig.DEFAULT_WORKER_COUNT))
                .setConfig(config)
                .setWorker(true);

        return Future.future(deployVerticle -> {
            vertx.deployVerticle(UrlCheckerVerticle.class.getName(), options, handler -> {
                if (handler.succeeded()) {
                    deployVerticle.complete();
                } else {
                    deployVerticle.fail("Failed to deploy [" + UrlCheckerVerticle.class.getName() + "] : " + handler.cause());
                }
            });
        });
    }

    private Future<Void> startServer() {
        return Future.future(startServer -> {
            Integer port = config.getInteger(ApplicationConfig.ENV_APPLICATION_PORT, ApplicationConfig.DEFAULT_APPLICATION_PORT);

            OpenAPI3RouterFactory.create(vertx, "webroot/openapi.yaml", handler -> {
                if (handler.succeeded()) {
                    OpenAPI3RouterFactory routerFactory = handler.result();

                    routerFactory.addHandlerByOperationId("checkUrl", this::checkUrl);

                    Router router = routerFactory.getRouter();
                    router.route("/*").handler(StaticHandler.create());

                    HttpServer server = vertx.createHttpServer(new HttpServerOptions().setPort(port));
                    server.requestHandler(router).listen();

                    log.info("Server successfully launched on port [{}]", port);
                    startServer.complete();
                } else {
                    startServer.fail(handler.cause());
                }
            });
        });
    }


    // passes urlCheckRequest to worker and handles response
    private void checkUrl(RoutingContext routingContext) {
        if (routingContext.getBodyAsString() != null) {
            // attempt decoding to detect malformed requests
            try {
                UrlCheckRequest request = Json.decodeValue(routingContext.getBodyAsString(), UrlCheckRequest.class);
                log.debug("Received valid {}", request.toString());

                vertx.eventBus().send(ApplicationConfig.MSG_URL_CHECK_ADDR, routingContext.getBodyAsString());

                routingContext.response()
                        .setStatusCode(202) // accepted
                        .putHeader("Content-Type", "application/json")
                        .end(routingContext.getBodyAsString());

            } catch (DecodeException e) {
                log.warn("Exception thrown while decoding request {}", routingContext.getBodyAsString(), e);
                routingContext.response().setStatusCode(400).end("Malformed request");
            }
        } else {
            routingContext.response().setStatusCode(400).end("No request body provided");
        }
    }
}


src/main/java/io/piveau/metrics/url_checker/UrlCheckResponse.java
package io.piveau.metrics.url_checker;

public class UrlCheckResponse {

    private String url;

    private Integer statusCode;

    private String mimeType;

    private String message;

    public UrlCheckResponse(String url) {
        this.url = url;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public Integer getStatusCode() {
        return statusCode;
    }

    public void setStatusCode(Integer statusCode) {
        this.statusCode = statusCode;
    }

    public String getMimeType() {
        return mimeType;
    }

    public void setMimeType(String mimeType) {
        this.mimeType = mimeType;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    @Override
    public String toString() {
        return "UrlCheckResponse{" +
                "url='" + url + '\'' +
                ", statusCode=" + statusCode +
                ", mimeType='" + mimeType + '\'' +
                ", message='" + message + '\'' +
                '}';
    }
}


src/main/java/io/piveau/metrics/url_checker/UrlCheckerVerticle.java
package io.piveau.metrics.url_checker;

import io.vertx.circuitbreaker.CircuitBreaker;
import io.vertx.circuitbreaker.CircuitBreakerOptions;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.CompositeFuture;
import io.vertx.core.Future;
import io.vertx.core.Promise;
import io.vertx.core.json.Json;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.WebClientOptions;
import io.vertx.ext.web.client.predicate.ResponsePredicate;
import org.apache.commons.validator.routines.UrlValidator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;


public class UrlCheckerVerticle extends AbstractVerticle {

    private static final Logger LOG = LoggerFactory.getLogger(UrlCheckerVerticle.class);

    private UrlValidator validator =
        new UrlValidator(UrlValidator.ALLOW_ALL_SCHEMES + UrlValidator.ALLOW_2_SLASHES + UrlValidator.ALLOW_LOCAL_URLS);

    private WebClient webClient;

    private CircuitBreakerOptions circuitBreakerOptions;


    @Override
    public void start(Promise<Void> startPromise) {

        int timeOutInMillis = config().getInteger(ApplicationConfig.ENV_TIMEOUT_MILLIS, ApplicationConfig.DEFAULT_TIMEOUT_IN_MILLIS);

        WebClientOptions clientOptions = new WebClientOptions()
            .setUserAgent(config().getString(ApplicationConfig.ENV_HTTP_USER_AGENT, ApplicationConfig.DEFAULT_HTTP_USER_AGENT))
            .setConnectTimeout(timeOutInMillis)
            .setIdleTimeout(timeOutInMillis)
            .setIdleTimeoutUnit(TimeUnit.MILLISECONDS)
            .setLogActivity(true);

        webClient = WebClient.create(vertx, clientOptions);

        circuitBreakerOptions = new CircuitBreakerOptions()
            .setMaxRetries(5)
            .setTimeout(10000)
            .setResetTimeout(60000);

        vertx.eventBus().<String>consumer(ApplicationConfig.MSG_URL_CHECK_ADDR, message -> {

            // decode request and check each url
            UrlCheckRequest urlCheckRequest = Json.decodeValue(message.body(), UrlCheckRequest.class);

            CompositeFuture.all(urlCheckRequest.getUrls().stream().map(this::checkUrl).collect(Collectors.toList())).setHandler(handler ->
                CircuitBreaker.create("callback-circuit-breaker", vertx, circuitBreakerOptions).execute(circuitBreakerFuture -> {
                    try {
                        webClient.postAbs(urlCheckRequest.getCallback())
                            .expect(ResponsePredicate.SC_OK)
                            .sendJson(handler.result().list(), response -> {
                                if (response.succeeded()) {
                                    circuitBreakerFuture.complete();
                                } else {
                                    circuitBreakerFuture.fail(response.cause());
                                }
                            });
                    } catch (Exception e) {
                        circuitBreakerFuture.fail(e);
                    }
                }).setHandler(circuitBreakerResult -> {
                    if (circuitBreakerResult.failed())
                        LOG.error("Failed to send URL callback to [{}]", urlCheckRequest.getCallback(), circuitBreakerResult.cause());
                })
            );
        });

        startPromise.complete();
    }

    private Future<UrlCheckResponse> checkUrl(String url) {

        return Future.future(checkUrl -> {
            if (url == null || !validator.isValid(encodeUrl(url))) {
                UrlCheckResponse urlCheckResponse = new UrlCheckResponse(url);
                urlCheckResponse.setStatusCode(config().getInteger(ApplicationConfig.ENV_INVALID_URL_STATUS, ApplicationConfig.DEFAULT_INVALID_STATUS_));
                urlCheckResponse.setMessage(config().getString(ApplicationConfig.ENV_INVALID_URL_MESSAGE, ApplicationConfig.DEFAULT_INVALID_URL_MESSAGE));

                LOG.debug("Encountered invalid URL: {}", urlCheckResponse.toString());
                checkUrl.complete(urlCheckResponse);
            } else {
                webClient.headAbs(url).send(handler -> {
                    UrlCheckResponse urlCheckResponse = new UrlCheckResponse(url);

                    if (handler.succeeded()) {
                        urlCheckResponse.setStatusCode(handler.result().statusCode());
                        urlCheckResponse.setMessage(handler.result().statusMessage());
                        urlCheckResponse.setMimeType(handler.result().getHeader("Content-Type"));
                    } else {
                        urlCheckResponse.setMessage(handler.cause().getMessage());
                        urlCheckResponse.setStatusCode(config().getInteger(ApplicationConfig.ENV_CONNECTION_EXCEPTION_STATUS, ApplicationConfig.DEFAULT_STATUS_CODE_EXCEPTION));
                    }

                    LOG.debug("Finished checking URL: {}", urlCheckResponse.toString());
                    checkUrl.complete(urlCheckResponse);
                });
            }
        });
    }

    private String encodeUrl(String url) {
        return url.replaceAll(" ", "%20");
    }
}


src/main/java/io/piveau/metrics/url_checker/ApplicationConfig.java
package io.piveau.metrics.url_checker;

final class ApplicationConfig {

    static final String MSG_URL_CHECK_ADDR = "url.check";

    static final String ENV_APPLICATION_PORT = "PORT";
    static final Integer DEFAULT_APPLICATION_PORT = 8080;

    static final String ENV_WORKER_COUNT = "WORKER_COUNT";
    static final Integer DEFAULT_WORKER_COUNT = 30;

    static final String ENV_TIMEOUT_MILLIS = "TIMEOUT_IN_MILLIS";
    static final Integer DEFAULT_TIMEOUT_IN_MILLIS = 10000;

    static final String ENV_CONNECTION_EXCEPTION_STATUS = "CONNECTION_EXCEPTION_STATUS";
    static final Integer DEFAULT_STATUS_CODE_EXCEPTION = 1100;

    static final String ENV_INVALID_URL_STATUS = "INVALID_URL_STATUS";
    static final Integer DEFAULT_INVALID_STATUS_ = 1300;

    static final String ENV_HTTP_USER_AGENT = "HTTP_USER_AGENT";
    static final String DEFAULT_HTTP_USER_AGENT = "Mozilla/5.0 (European Data Portal) Gecko/20100101 Firefox/40.1";

    static final String ENV_INVALID_URL_MESSAGE = "INVALID_URL_MESSAGE";
    static final String DEFAULT_INVALID_URL_MESSAGE = "Invalid URL";
}
